sparse/mlir_backend/_constructors.py

```diff
@@ -0,0 +1,259 @@
+            if DEBUG:
+                (SCRIPT_PATH / "csr_module_opt.mlir").write_text(str(module))
+
+        module = mlir.execution_engine.ExecutionEngine(module, opt_level=2, shared_libs=[MLIR_C_RUNNER_UTILS])
+        return (module, csr_shaped)
+
+    @classmethod
+    def assemble(cls, module: ir.Module, arr: sps.csr_array) -> ctypes.c_void_p:
+        out = ctypes.c_void_p()
+        module.invoke(
+            "assemble",
+            ctypes.pointer(ctypes.pointer(MemrefIdx_1D.from_numpy(arr.indptr))),
+            ctypes.pointer(ctypes.pointer(MemrefIdx_1D.from_numpy(arr.indices))),
+            ctypes.pointer(ctypes.pointer(MemrefF64_1D.from_numpy(arr.data))),
+            ctypes.pointer(out),
+        )
+        return out
+
+    @classmethod
+    def disassemble(cls, module: ir.Module, ptr: ctypes.c_void_p) -> sps.csr_array:
+        class Csr(ctypes.Structure):
+            _fields_ = [
+                ("data", MemrefF64_1D),
+                ("pos", MemrefIdx_1D),
+                ("crd", MemrefIdx_1D),
+                ("data_len", np.ctypeslib.c_intp),
+                ("pos_len", np.ctypeslib.c_intp),
+                ("crd_len", np.ctypeslib.c_intp),
+                ("shape_x", np.ctypeslib.c_intp),
+                ("shape_y", np.ctypeslib.c_intp),
+            ]
+
+            def to_sps(self) -> sps.csr_array:
+                pos = self.pos.to_numpy()[: self.pos_len]
+                crd = self.crd.to_numpy()[: self.crd_len]
+                data = self.data.to_numpy()[: self.data_len]
+                return sps.csr_array((data, crd, pos), shape=(self.shape_x, self.shape_y))
+
+        arr = Csr()
+        module.invoke(
+            "disassemble",
+            ctypes.pointer(ctypes.pointer(arr)),
+            ctypes.pointer(ptr),
+        )
+        return arr.to_sps()
+
+
+def _is_scipy_sparse_obj(x) -> bool:
+    return hasattr(x, "__module__") and x.__module__.startswith("scipy.sparse")
+
+
+def _is_numpy_obj(x) -> bool:
+    return isinstance(x, np.ndarray)
+
+
+def asarray(obj) -> Tensor:
+    # TODO: discover obj's dtype
+    values_dtype = Float64
+    index_dtype = Index
+
+    # TODO: support other scipy formats
+    if _is_scipy_sparse_obj(obj):
+        format_class = CSRFormat
+    elif _is_numpy_obj(obj):
+        format_class = DenseFormat
+    else:
+        raise Exception(f"{type(obj)} not supported.")
+
+    # TODO: support proper caching
+    if hash(obj.shape) in format_class.modules:
+        module, tensor_type = format_class.modules[hash(obj.shape)]
+    else:
+        module, tensor_type = format_class.get_module(obj.shape, values_dtype, index_dtype)
+        format_class.modules[hash(obj.shape)] = module, tensor_type
+
+    assembled_obj = format_class.assemble(module, obj)
+    return Tensor(assembled_obj, module, tensor_type, format_class.disassemble, values_dtype, index_dtype)
+
+
+def free(x: Tensor) -> None:
+    x.module.invoke("free", ctypes.pointer(x.obj))
```

Comment:
Weirdly calling `invoke("free", ...` causes a seg fault for me, I commented `__del__` out for now.

Reply:
I think if we assemble a tensor, it re-uses the memory, but disassembling doesn't. So if we assemble the tensor, we should set `t.owned = True` and otherwise `False`. Then we should only call `free` if the tensor is owned.

Reply:
Ouch. Now I understand. We need to hold a reference to the tensor in the SciPy array, look into the `weakref` module.

Reply:
Ok, in the last PR I added `__del__` method to the Tensor class and removed `free_memory` call. I checked with `psutil` that constructing Tensors and converting to SciPy objects doesn't cause a memory leak. I think that `MemrefType` classes don't own any data so it shouldn't free it.